daemon/snapper/snapper.go

```diff
@@ -257,13 +262,58 @@ func snapshot(a args, u updater) state {
 			progress.state = SnapStarted
 		})
 
-		l.Debug("create snapshot")
-		err := zfs.ZFSSnapshot(fs, snapname, false)
-		if err != nil {
+		hookEnv := map[string]string{
+			"ZREPL_HOOKTYPE": "pre",
+			"ZREPL_PHASE": "Snapshotting",
+			"ZREPL_FS": fs.ToString(),
+			"ZREPL_SNAPNAME": snapname,
+		}
+		var err error
+		var preHookErr error
+		var doneAt time.Time
+		if (a.hooks.Pre != "") {
+			l.Debug("pre-snapshot hook")
+			preHookErr = hooks.RunHookCommand(
+				a.ctx,
+				a.hookDir,
+				a.hooks.Pre,
+				hookEnv,
+				a.hooks.Timeout,
+			)
+			if preHookErr != nil {
+				l.WithError(preHookErr).Error("cannot run pre-snapshot hook")
+			}
+		}
+
+		if preHookErr == nil || a.hooks.Keep {
+			l.Debug("create snapshot")
+			err = zfs.ZFSSnapshot(fs, snapname, false)
+			if err != nil {
+				hadErr = true
+				l.WithError(err).Error("cannot create snapshot")
+			}
+			doneAt = time.Now()
+
+			if !hadErr {
+                hookEnv["ZREPL_HOOKTYPE"] = "post"
```

Comment:
ugh, not so pretty to mutate hookEnv here. factory function à la `func makeHookEnv(edge "pre"|"post")`?
Could also use it to generate the (proposed) argument struct to hooks.RunHookCommand.